Feat/uu list 4

RoordinkFrank opened the pull request.

EdoStorm96 committed:
feat: add vue3 to settings.py

EdoStorm96 committed:
feat: create ProposalApiView

EdoStorm96 committed:
feat: redo MyProposalsView using UU-list

EdoStorm96 committed:
fix: remove unused context

RoordinkFrank committed:
Continued implementation
Todo list is on github and/or in the comments of this commit

RoordinkFrank commented on main/serializers.py:
```diff
@@ -8,3 +8,12 @@ class Meta:
+
+
+# redundant text with above, is it possible to override thhe fields?
+class UserMinimalSerializer(serializers.ModelSerializer):
```
I do prefer to be able to write this comment before a draft pull request (or commit). So if anyone knows how to do that.

Back to the original comment. The minimal serializer gives three value, fullname for display, first and last name for the filter. I suspect the filter getś it's data from somehwere else and is unneccesary.

RoordinkFrank committed:
moved logic from serializer to proposal_actions
A few links have been changed to redirect to the new view
review_pk added to proposal. The serializer needs a model.variable directly.

RoordinkFrank committed:
removed make decision from general view
clarified archive actions
cleaned DDVAction field a bit

RoordinkFrank commented on proposals/api/serializers.py:
```diff
@@ -76,31 +88,126 @@ class Meta:
+
+    @staticmethod
+    def get_state(proposal: Proposal):
+        # state = status with decisions added so this is still lacking
```
status displayed still needs more, decision of the ethics commision should be displayed here as well.

RoordinkFrank commented on proposals/views/proposal_views.py:
```diff
@@ -86,18 +96,76 @@ def get_context_data(self, **kwargs):
+            # compared to original data shown.
+            label="date submitted",
+        ),
+        DDVDate(  # DDVDate is an option, however the default for their empty dat is 1970 instead of empty.
```
DDVDate has 1970 as a default option when there is an empty field. That is different with current behaviour.

RoordinkFrank committed:
diff action is working

RoordinkFrank committed:
Revision now uses is_revisionable. Revision button are shown in the same places as develop.

RoordinkFrank committed:
naar volgende stap correcte translatation
made get_state_or_decision field. Stat is now working correctly

RoordinkFrank committed:
Dates now show in date format instead of isoformat.

RoordinkFrank committed:
fixed the filter showing two times.

RoordinkFrank committed:
Date modified now works correctly

RoordinkFrank committed:
Added my supervised
removed supervised applications from my_applications
decide button not yet correct
new revision keeps appearing in my supervised while that should not happen.

RoordinkFrank committed:
Supervise action is on the correct proposals.
The action itself does not work yet.

RoordinkFrank commented on proposals/models.py:
```diff
@@ -776,11 +786,18 @@ def supervisor_decision(self):
         from reviews.models import Review
 
         return Review.objects.filter(proposal=self).last()
 
```
the serializer seems to want a direct field of the model. And a field of a field is not accepted. Still this is not a great solution as it clutters the proposal model even more than it already is.
I could make a separate model matching the serializer where there are methods like these. That would remove this from proposal. I think it may end up slightly better but not ideal either.

RoordinkFrank committed:
supervise extra info added, does nnot yet match orignal gui extra info of supervise

RoordinkFrank committed:
supervisor decision button now works

RoordinkFrank committed:
cleanup
moved serializer logic to DDVProposalSerializer

RoordinkFrank committed:
added delete action to supervisor

RoordinkFrank committed:
separator for delete added and working

RoordinkFrank committed:
Added practise applications for DDV
Removed redunancy in practise applications and supervised applications
Removed unused menu buttons

RoordinkFrank committed:
black .

RoordinkFrank commented on proposals/api/views.py:
```diff
@@ -2,16 +2,91 @@
+
+class MyPracticeApiView(ProposalApiView):
+    """Gets all practise applications including supervisor."""
+
```
# practise proposals for supervisor currently do appear in MySupervised to match the action buttons.
    # Uses probably do expect it in this view but I am unsure how to implement that.

RoordinkFrank commented on proposals/api/serializers.py:
```diff
@@ -99,8 +112,174 @@ class Meta:
+        return ""
+
+    # logic works but unsure if it should be in the serializer
+    @staticmethod
```
logic works but unsure if it should be in the serializer

RoordinkFrank replied:
Moved logic to proposal

RoordinkFrank commented on proposals/api/serializers.py:
```diff
@@ -99,8 +112,174 @@ class Meta:
+
+    action_make_revision = DDVLinkField(
+        text=_("Maak revisie"),
+        link="proposals:copy",  # there is also a copy revison and copy amendment,
```
there is also a copy revison and copy amendment,
         not sure when those two are supposed to be shown

RoordinkFrank commented on proposals/utils/proposal_actions.py:
```diff
@@ -0,0 +1,91 @@
+             <img src="{{ img_hide }}" title="Verbergen">
+               <img src="{{ img_add }}" title="Toevoegen">
+"""
+
```
import datetime

# actiontree, implementation in proosal_actions.py
"""
    {% if modifiable %}
        <img src="{{ img_next }}" title="Naar volgende stap">
        <img src="{{ img_diff }}" title="Toon verschillen">
        <img src="{{ img_delete }}" title="Verwijderen">
     {% if submitted %} 
        <img src="{{ img_pdf }}" title="Inzien">
        <img src="{{ img_revise }}" title="Maak revisie">
      {% if supervised %}
          <img src="{{ img_decide }}" title="Beslissen">
       {% if is_secretary %}
             <img src="{{ img_hide }}" title="Verbergen">
               <img src="{{ img_add }}" title="Toevoegen">
"""

RoordinkFrank replied:
ah so we have a templates/proposals/proposal_list.html and a templates/proposals/vue_templates/proposal_list.html

RoordinkFrank commented on proposals/utils/proposal_actions.py:
```diff
@@ -0,0 +1,91 @@
+"""
+
+
+# Does not check user permissions as of now
```
Does not check user permissions as of now

RoordinkFrank committed:
removed comments and added them to the pullrequest

RoordinkFrank commented on proposals/views/proposal_views.py:
```diff
@@ -86,21 +96,115 @@ def get_context_data(self, **kwargs):
+    #        label=_("Acties"),
+    #    ),
+    # )
+    columns = [
```
# I do want to do something like below to avoid duplicate DDVcolumns but I can not get it to work
    # the subclass does not seem to know the superclass variable if they are class variables.
    # I have a simular problem in the serializer, if there is someone who might know a solution please say so.
    # columns.append(
    #    DDVActions(
    #        field="my_supervised_actions",
    #        label=_("Acties"),
    #    ),
    # )

RoordinkFrank commented on proposals/models.py:
```diff
@@ -622,6 +621,14 @@ class PracticeReasons(models.IntegerChoices):
+    def get_applicants_names(self):
+        users = ""
+        for user in self.applicants.all():
+            if users != "":
```
.all() has a yellow error. Unsolved attribute. It works fine but I have no idea how to remove the error.

RoordinkFrank commented on proposals/utils/proposal_actions.py:
```diff
@@ -0,0 +1,66 @@
+    @staticmethod
+    def action_allowed_show_difference(proposal: Proposal):
+        # if_modifiable return true
+        if proposal.is_revision:  # amendment_or_revision is also an option?
```
amendment_or_revision is also an option?

RoordinkFrank commented on proposals/utils/proposal_actions.py:
```diff
@@ -0,0 +1,66 @@
+    # hid and add to archive may not belong here, but I will not remove them without confirmation that they do not belong
+    # in a proposal view.
+    @staticmethod
+    def action_allowed_hide_from_archive(proposal: Proposal):
```
Hide and show will be implemented during the archive DDL view.

RoordinkFrank committed:
Removed userminserializer, it was no longer used.
Removed more comments and added them back as github comments

RoordinkFrank commented on proposals/views/proposal_views.py:
```diff
@@ -86,21 +96,115 @@ def get_context_data(self, **kwargs):
+            label=_("Status"),
+        ),
+        DDVString(
+            field="date_modified",  # in the old version this is either laatst bijgewerkt or besloten op.
```
in the old version this is either laatst bijgewerkt or besloten op.
That last interface part is now lost.

RoordinkFrank committed:
removed even more comments.

RoordinkFrank commented on proposals/utils/proposal_actions.py:
```diff
@@ -0,0 +1,50 @@
+    def action_allowed_hide_from_archive(proposal: Proposal):
+        check = lambda o: o.status == Proposal.Statuses.SUBMITTED
+        return check(proposal)
+
```
lamda logic is left intact so it can be ported back to the serializer if ProposalActions gets removed.

EdoStorm96 commented on proposals/utils/proposal_actions.py:
```diff
@@ -0,0 +1,66 @@
+    def action_allowed_view_pdf(proposal: Proposal):
+        # if_submitted return true
+        # does seem to be always available, why is there a submitted requirement?
+        return proposal.status == Proposal.Statuses.SUBMITTED
```
The pdf might be always availabe ... But we only link to it after it gets submitted. I think that makes sense so let's keep it that way.

EdoStorm96 commented on proposals/utils/proposal_actions.py:
```diff
@@ -0,0 +1,50 @@
+import datetime
+
+
+class ProposalActions:
```
I do see how this is a problem ... But I think we can overcome this by carefully constructing querysets and splitting out in a few lists, like you did with my_proposals and my_supervised. Because we can access the user in the api views, we can ensure that only relevant proposals appear in a certain list. Then we can only make actions that should only be available to a specific user, only available in the specific list where we know only that specific user can come. I think taken together, this will create a solid system.

EdoStorm96 reviewed (commented):
Ok, so really good work! Massive props for figuring all of this out by yourself. That must have been challenging. I think it is looking really snazzy! I think really the main stuff is done, but we are just missing some functionality and polish. But overall this is really on the right track!

I tried to be as thorough as possible with my review, but I might have missed some stuff. I just found a lot of little things which is to be expected with such a large PR. I hope you find my comments useful. But in general you can keep going in this direction I think!

I have one general comment: ensure al references to removed list views are also removed eg. proposal_submitted.html.

It will probably take quite a while for this to be fully finished, as it is a really big feature, but do not be discouraged by that. We will need to iterate over this a few times, before it gets perfect (also on the review side) and hopefully @miggol can take a look as well. But in general this is going in the right direction!

RoordinkFrank commented on proposals/models.py:
```diff
@@ -741,6 +748,7 @@ def amendment_or_revision(self):
         if self.is_revision and self.parent:
             return _("Amendement") if self.parent.status_review else _("Revisie")
 
+    @property
```
The added @property? Serializer does not recognize methods and gives back a yellow error without it.

RoordinkFrank commented on proposals/api/views.py:
```diff
@@ -2,16 +2,88 @@
+        label="Status",
+        field_name="status",
+        choices=status_choices,
+    )
```
Definitely looks cleaner already.

RoordinkFrank commented on proposals/api/serializers.py:
```diff
@@ -99,8 +112,174 @@ class Meta:
+    info = serializers.SerializerMethodField()
+
+    # I am having trouble finding out what info best to display to the supervisor
+    @staticmethod
```
done

RoordinkFrank committed:
renamed state_or_decision to detailed_state, still not ideal but slightly more accurate.
moved to proposal as the logic is about the model.
changed supervisor_decision_pk after @edo's suggestion.

RoordinkFrank commented on reviews/utils/review_actions.py:
```diff
@@ -9,6 +9,7 @@
 import datetime
 
 
+# revviewactions are here
```
done

RoordinkFrank committed:
made status filter cleaner after edo's suggestion.
proposal.get_applicants_names is slightly cleaner after edo's comment.
reversed compared documents url.

RoordinkFrank committed:
made status filter cleaner after edo's suggestion.
proposal.get_applicants_names is slightly cleaner after edo's comment.
reversed compared documents url.
fixed when pdf's are shown.

RoordinkFrank commented on proposals/api/serializers.py:
```diff
@@ -99,8 +112,157 @@ class Meta:
+
+    # DDVDate returns 1970 by default so we have to use a DDVString and format the date ourselves
+    # incoming format: 2025-04-24T11:59:45.583305+02:00
+    @staticmethod
```
<div class="ufl-undertitle-line">
                <span v-if="proposal.latest_review && proposal.latest_review.continuation == context.review.REVISION">
                    {% trans "Besloten op" %}:
                    {% verbatim %}
                    {{ proposal.date_reviewed | date("DD-MM-YYYY, HH:mm") }}
                    {% endverbatim %}
                </span>
                <span
                    v-else-if="proposal.date_confirmed"
                >
                    {% trans "Besloten op" %}:
                    {% verbatim %}
                    {{ proposal.date_confirmed | date("DD-MM-YYYY, HH:mm") }}
                    {% endverbatim %}
                </span>
                <span v-else>
                    {% trans "Laatst bijgewerkt" %}:
                    {% verbatim %}
                    {{ proposal.date_modified | date("DD-MM-YYYY, HH:mm") }}
                    {% endverbatim %}
                </span>
            </div>

Basicly fusing these three, we only have place for one label though.

RoordinkFrank commented on proposals/api/serializers.py:
```diff
@@ -99,8 +112,157 @@ class Meta:
+        if proposal.date_modified is not None:
+            return proposal.date_modified.date()
+        return ""
+
```
date_submitted is not date submitted to supervisor but date submitted to commission?
the wording of date_submitted seems to be wrong.

RoordinkFrank committed:
pre-secretary and archive commit.

RoordinkFrank committed:
removed secretary stuff, that is separate branch now.

RoordinkFrank committed:
removed duplication of columns in proposal_view.py.

RoordinkFrank committed:
fixed tests

RoordinkFrank committed:
black .

RoordinkFrank committed:
small comment change

miggol reviewed (commented):
This is really cool! Congrats on getting this to work. There's still a way to go. I have some comments to start:

1. I don't see much utility in filtering by status. I think the left-hand checkbox filters should be "supervised", "draft", "processed", etc. This can supplant the need for all the separate views we have today.
2. There are too many columns displaying too much text. Text density is high while information density is low. This can be fixed by having a detail view, or having a row "expand" when clicked like with the current lsits. Or perhaps use icons or colours to represent certain information.
3. I'm a bit conflicted about the actions implementation. I like that they work and are understandable. But they are hard to reuse and maintain in this way. Perhaps we should have a talk about this in person, I'm keen to hear your opinion on doing it either way.

UU-list-4 will not be ready for merging into develop anytime soon, this is more of a long term feature. I would recommend going further as follows:

* This branch becomes the new base for uu-list development
* Further changes come from PRs into this branch, which can come from anyone and hopefully be smaller changesets. I see you've already started on this!
* Perhaps we can put this branch on the test server. Edo knows how to do this. This does require that it is kept to a certain degree of usability.
* I think that when we get close to merging this, we will want a fresh PR without dusty old comments. So IMO this PR can be closed while we keep working on the branch. But I don't mind keeping it either.

RoordinkFrank commented on proposals/api/serializers.py:
```diff
@@ -99,8 +112,160 @@ class Meta:
+        link="proposals:pdf",
+        link_attr="pk",
+        new_tab=True,
+        check=lambda proposal: ProposalActions.action_allowed_view_pdf(proposal),
```
That makes way more sense. It's just that I don´t know how. My tries didn´t work so I did leave it as it is.
